relayer/chains/substrate/event_parser.go

```diff
@@ -1,29 +1,478 @@
 package substrate
 
 import (
+	"fmt"
+	"strconv"
+
 	rpcclienttypes "github.com/ComposableFi/go-substrate-rpc-client/v4/types"
+	beefyclienttypes "github.com/ComposableFi/ics11-beefy/types"
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
-	"github.com/cosmos/relayer/v2/relayer/processor"
 	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
-// ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
-func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
-	//TODO implement me
-	panic("implement me")
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	parseAttrs(log *zap.Logger, attrs interface{})
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
+type ibcEventQueryItem (map[string]interface{})
+
+// ibcMessagesFromEvents parses all events of a certain height to find IBC messages
+func (scp *SubstrateChainProcessor) ibcMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64) (messages []ibcMessage) {
+	packetAccumulator := make(map[ibcPacketKey]*packetInfo)
+	for i := 0; i < len(ibcEvents); i++ {
+		var info ibcMessageInfo
+		var eventType string
+
+		for eType, data := range ibcEvents[i] {
+
+			switch eType {
+
+			case CreateClient, UpgradeClient, ClientMisbehaviour:
+
+				ce := new(clientInfo)
+				ce.parseAttrs(scp.log, data)
+				info = ce
+
+				eventType = intoIBCEventType(eType)
+
+			case UpdateClient:
+
+				ce := new(clientUpdateInfo)
+				ce.parseAttrs(scp.log, data)
+				info = ce
+
+				eventType = intoIBCEventType(eType)
+
+			case OpenInitConnection, OpenTryConnection, OpenAckConnection, OpenConfirmConnection:
+
+				ce := &connectionInfo{Height: height}
+				ce.parseAttrs(scp.log, data)
+				info = ce
+
+				eventType = eType
+
+			case OpenInitChannel, OpenTryChannel, OpenAckChannel, OpenConfirmChannel, CloseInitChannel, CloseConfirmChannel:
+				ce := &channelInfo{Height: height}
+				ce.parseAttrs(scp.log, data)
+				info = ce
+
+				eventType = intoIBCEventType(eType)
+
+			case SendPacket, ReceivePacket, WriteAcknowledgement, AcknowledgePacket, TimeoutPacket, TimeoutOnClosePacket:
+
+				accumKey := genAccumKey(data)
+
+				_, exists := packetAccumulator[accumKey]
+				if !exists {
+					packetAccumulator[accumKey] = &packetInfo{Height: height}
+				}
+
+				packetAccumulator[accumKey].parseAttrs(scp.log, data)
+
+				info = packetAccumulator[accumKey]
```

Comment:
rather than setting `info` here, since the accumulation might not be done yet for the packetInfo, we can remove this line, then iterate over the `packetAccumulator` map outside of the for loop (down on line 103) and append the final accumulated packetInfos to `messages` there.

Reply:
In the substrate chain query here, we are able to get events of a certain height, this means that it is possible to rewrite the `info` by other irrelevant event info in client, connection, and channel events as well. so I think we should store all of them in separate lists and loop through them after for loop finish. I think in the current approach we will lost most of events and create last event message in append to messages list.

Reply:
> I think we should store all of them in separate lists

The `packetAccumulator` map here is already used for this to store packetInfo for unique channel/sequence.

> ... and loop through them after for loop finish.

this can be iterating over the `packetAccumulator` on line 103 to append to `messages`, and removing line 81.

Reply:
I am talking about other types of events than packet -related events, what about them? There is no guarantee that exist one and only one of these types. For example events may be consist of multiple CreateClient, but this approach assumes that there is only one CreateClient which is set into ClientInfo.

Reply:
the only reason for the packetAccumulator in the cosmos implementation is to accumulate the ack bytes from the `write_acknowledgement` event with `recv_packet` event. Other than that, all other IBC messages can be considered as 1-1 with the event. This brings up a good point though, this switch case only needs to handle the `ReceivePacket` and `WriteAcknowledgement`. There can be another switch case added for `SendPacket, AcknowledgePacket, TimeoutPacket, TimeoutOnClosePacket` which can do the single `parseAttrs` and set `info` like the other switch cases.

Reply:
Yes, And also in substrate list of events, there is no guarantee for 1-1 event and message mapping, So I need to redesign this according to `ibc_events` query endpoint of composable chain

Reply:
Done

Reply:
> For example events may be consist of multiple CreateClient, but this approach assumes that there is only one CreateClient which is set into ClientInfo.

Yes, `var info ibcMessageInfo` should move to line 36 and the bracket on line 90 should move to line 101 so that the `info` nil check and `messages` append are on each event.

Reply:
I see that info is now scoped inside the events loop in the latest, that looks good.